paymaster/test/tests.ts

```diff
@@ -0,0 +1,124 @@
+import { ethers } from 'ethers';
+import dotenv from 'dotenv';
+import axios from 'axios';
+import { UserOperation } from '../src/rpcMethods';
+import { simpleAccountAbi, simpleAccountFactoryAbi, paymasterAbi, entrypointAbi } from '../src/abis';
+dotenv.config();
+
+async function main() {
+    const provider = new ethers.JsonRpcProvider(process.env.RPC_URL as string);
+    const signer = new ethers.Wallet(process.env.PRIVATE_KEY as string, provider);
+
+    // Paymaster first needs to have funds in Entrypoint contract
+    const paymasterAddress = process.env.PAYMASTER_ADDRESS as string;
+    const paymaster = new ethers.Contract(
+        paymasterAddress,
+        paymasterAbi,
+        signer
+    );
+    await (await paymaster.deposit({ value: ethers.parseEther('3')})).wait();
+    await (await paymaster.addStake(3000, { value: ethers.parseEther('3')})).wait();
+
+    // Generate initcode
+    const simpleAccountFactoryAddress = process.env.SIMPLE_ACCOUNT_FACTORY_ADDRESS as string;
+    const simpleAccountFactory = new ethers.Contract(
+        simpleAccountFactoryAddress,
+        simpleAccountFactoryAbi,
+        provider
+    )
+    const initCode = ethers.concat([
+        simpleAccountFactoryAddress,
+        simpleAccountFactory.interface.encodeFunctionData("createAccount", [signer.address, 0]) // set salt as 0
+    ]);
+
+    // Generate calldata to execute a transaction
+    const simpleAccountAddress = await simpleAccountFactory.getFunction("getAddress")(signer.address, 0);
+    const simpleAccount = new ethers.Contract(
+        simpleAccountAddress,
+        simpleAccountAbi,
+        provider
+    )
+    const to = "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045" // vitalik
+    const value = 0
+    const data = "0x68656c6c6f" // "hello" encoded to utf-8 bytes
+    const callData = simpleAccount.interface.encodeFunctionData("execute", [to, value, data]);
+
+    // Construct UserOp
+    let userOp: UserOperation = {
+        sender: simpleAccountAddress,
+        nonce: Number(await paymaster.senderNonce(simpleAccountAddress)),
+        initCode: initCode,
+        callData: callData,
+        callGasLimit: ethers.toBeHex(3_000_000), // hardcode it for now at a high value,
+        verificationGasLimit: ethers.toBeHex(3_000_000), // hardcode it for now at a high value,
+        preVerificationGas: ethers.toBeHex(2_000_000), // hardcode it for now at a high value,
+        maxFeePerGas: ethers.toBeHex(2e9),
+        maxPriorityFeePerGas: ethers.toBeHex(1e9),
+        paymasterAndData: ethers.concat([
+            paymasterAddress,
+            '0x' + '00'.repeat(64),
+            '0x' + '00'.repeat(65)
```

Comment:
65?
I see elsewhere that's two `uint48` (`validAfter`, `validUntil`). I know ABI might extend them to 32 bytes but why the extra byte then?

Reply:
65 is for the signature of the paymaster server. `validAfter` and `validUntil` both takes 32bytes each, which is the `'0x' + '00'.repeat(64)`.